server/graphql/types/src/types/currency.rs

```diff
@@ -0,0 +1,144 @@
+use async_graphql::*;
+use chrono::NaiveDate;
+use graphql_core::generic_filters::EqualFilterStringInput;
+use graphql_core::simple_generic_errors::NodeError;
+use repository::{
+    Currency, CurrencyFilter, CurrencyRow, CurrencySort, CurrencySortField, EqualFilter,
+};
+use service::{usize_to_u32, ListResult};
+
+#[derive(Enum, Copy, Clone, PartialEq, Eq)]
+#[graphql(rename_items = "camelCase")]
+pub enum CurrencySortFieldInput {
+    Id,
+    CurrencyCode,
+}
+#[derive(InputObject)]
+pub struct CurrencySortInput {
+    /// Sort query result by `key`
+    key: CurrencySortFieldInput,
+    /// Sort query result is sorted descending or ascending (if not provided the default is
+    /// ascending)
+    desc: Option<bool>,
+}
+
+#[derive(InputObject, Clone)]
+pub struct CurrencyFilterInput {
+    pub id: Option<EqualFilterStringInput>,
+    pub is_home_currency: Option<bool>,
+    pub is_active: Option<bool>,
+}
+
+impl CurrencyFilterInput {
+    pub fn to_domain(self) -> CurrencyFilter {
+        let CurrencyFilterInput {
+            id,
+            is_home_currency,
+            is_active,
+        } = self;
+
+        CurrencyFilter {
+            id: id.map(EqualFilter::from),
+            is_home_currency,
+            is_active,
+        }
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub struct CurrencyNode {
+    pub currency: Currency,
+}
+
+#[derive(SimpleObject)]
+pub struct CurrencyConnector {
+    total_count: u32,
+    nodes: Vec<CurrencyNode>,
+}
+
+#[Object]
+impl CurrencyNode {
+    pub async fn id(&self) -> &str {
+        &self.row().id
+    }
+
+    pub async fn code(&self) -> &str {
+        &self.row().code
+    }
+
+    pub async fn rate(&self) -> f64 {
+        self.row().rate
+    }
+
+    pub async fn is_home_currency(&self) -> bool {
+        self.row().is_home_currency
+    }
+
+    pub async fn date_updated(&self) -> Option<NaiveDate> {
+        self.row().date_updated
+    }
+
+    pub async fn is_active(&self) -> bool {
+        self.row().is_active
+    }
+}
+
+#[derive(Union)]
+pub enum CurrenciesResponse {
```

Comment:
These query endpoint specific types (response, sort and filter input) should be specific to the `currency`/`currencies` endpoint so they don't need to be here but can go into the query endpoint file.

What about the CurrencyNode? Do you plan to use it in other graphql crates? Otherwise, I would also keep them in the currency query file...

The types crate was just for types that need to be shared between graphql crates...

Reply:
Ah right. Yeah, I don't think it'll be used anywhere else. 

Thanks for the review Clemens~ all fixes can be found here [d6ff0ea](https://github.com/msupply-foundation/open-msupply/pull/2810/commits/d6ff0ea1aad359e9541083e09764b91014a4143c)